File: app/frontend/src/game/visuals.ts
export type ArtCategory = 'tavern' | 'wilds' | 'harbor' | 'ruins' | 'abstract' | 'portrait' | 'traveller' | 'weapon' | 'armor' | 'potion' | 'item';
export interface ArtSlot { asset_id?: string | null; builtin?: string | null; fallback?: ArtCategory | null; focal_x?: number; focal_y?: number }
export interface VisualAsset { name: string; source: 'upload' | 'generated'; data: string; generation_id?: string | null }
export interface ThemePack { schema_version?: 1; id: string; name: string; preset: 'parchment' | 'midnight' | 'neutral'; heading_font?: 'serif' | 'sans'; frame?: 'ornate' | 'simple'; illustrated?: boolean; assets?: Record<string, VisualAsset>; defaults?: Partial<Record<ArtCategory, ArtSlot>> }
export interface ModuleVisuals { schema_version?: 1; theme?: ThemePack; assets?: Record<string, VisualAsset>; scenes?: Record<string, ArtSlot>; characters?: Record<string, ArtSlot>; items?: Record<string, ArtSlot>; map?: Record<string, ArtSlot>; map_layout?: Record<string, {x:number; y:number; label?:string | null}>; player?: ArtSlot | null; cover?: ArtSlot | null }
export const themes: ThemePack[] = [
  { id: 'western-journal', name: '西境旅行志', preset: 'parchment', heading_font: 'serif', frame: 'ornate', illustrated: true },
  { id: 'moonlit-chronicle', name: '月下编年史', preset: 'midnight', heading_font: 'serif', frame: 'ornate', illustrated: true },
  { id: 'universal-atlas', name: '通用星图', preset: 'neutral', heading_font: 'sans', frame: 'simple', illustrated: false },
];
export const categoryNames: Record<ArtCategory, string> = { tavern: '室内 / 酒馆', wilds: '旷野 / 林地', harbor: '聚落 / 水岸', ruins: '遗迹 / 地下', abstract: '通用抽象背景', traveller: '冒险者', portrait: '人物徽记', weapon: '武器', armor: '防具', potion: '药剂', item: '物品 / 线索' };
export function themeOf(visuals?: ModuleVisuals | null) { return visuals?.theme || themes[0]; }
export function sceneCategory(name = ''): ArtCategory {
  if (/酒馆|客栈|室|屋|馆|tavern|inn\b/i.test(name)) return 'tavern';
  if (/遗迹|地下|墓|洞|crypt|dungeon|ruin/i.test(name)) return 'ruins';
  if (/港|渡|码头|河|海|城|镇|村|harbor|ferry|town|city/i.test(name)) return 'harbor';
  return 'wilds';
}
export function itemCategory(type = ''): ArtCategory { return type === 'weapon' ? 'weapon' : type === 'armor' ? 'armor' : type === 'consumable' ? 'potion' : 'item'; }
const icons: Record<string, string> = {
  portrait: '<circle cx="50" cy="34" r="13"/><path d="M22 80Q24 54 50 54Q76 54 78 80Z"/>',
  traveller: '<path d="M25 82L32 35Q50 7 68 35L75 82ZM32 35L50 50L68 35M50 50V82"/>',
  weapon: '<path d="M25 76L72 20L82 18L82 29L36 81M24 58L47 79M18 85L29 73"/>',
  armor: '<path d="M24 20L40 16Q50 33 60 16L76 20L86 41L71 49L71 84L29 84L29 49L14 41ZM39 45H61M50 35V70"/>',
  potion: '<path d="M40 18H60V38Q80 54 78 74Q75 87 50 87Q25 87 22 74Q20 54 40 38ZM35 18H65M28 62H72"/>',
  item: '<path d="M22 23H71L80 32V78H22ZM71 23V34H80M34 43H66M34 55H61M34 67H56"/>',
};
export function defaultArt(category: ArtCategory, illustrated = true): string {
  if (illustrated && ['tavern', 'wilds', 'harbor', 'ruins', 'traveller'].includes(category)) return `/art/western/${category}.png`;
  if (['tavern', 'wilds', 'harbor', 'ruins', 'abstract'].includes(category)) return '/art/atlas.svg';
  return `data:image/svg+xml,${encodeURIComponent(`<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100"><rect width="100" height="100" fill="#263c3a"/><circle cx="50" cy="50" r="46" fill="none" stroke="#7e8060"/><g fill="none" stroke="#e2cb9a" stroke-width="2.5" stroke-linejoin="round">${icons[category] || icons.item}</g></svg>`)}`;
}
export function builtinArt(builtin?: string | null, illustrated = true): string | null {
  return illustrated && builtin && ['tavern','wilds','harbor','ruins','traveller','village-square','marcus','ayla','hooded-merchant', 'dungeon-gate', 'dungeon-passage', 'ancient-temple', 'ancient-vault'].includes(builtin) ? `/art/western/${builtin}.png` : null;
}
export function resolveArt(visuals: ModuleVisuals | null | undefined, slot: ArtSlot | null | undefined, fallback: ArtCategory) {
  const theme = themeOf(visuals), category = slot?.fallback || fallback;
  const inherited = theme.defaults?.[category];
  const custom = slot?.asset_id ? visuals?.assets?.[slot.asset_id] : undefined;
  const shared = inherited?.asset_id ? theme.assets?.[inherited.asset_id] : undefined;
  const asset = custom || shared;
  const safe = asset?.data && /^data:image\/(png|jpeg);base64,[A-Za-z0-9+/=]+$/.test(asset.data) && asset.data.length <= 240000;
  const builtin = slot?.builtin || inherited?.builtin;
  const curated = builtinArt(builtin, theme.illustrated !== false);
  const focus = custom || (!shared && slot?.builtin) ? slot : inherited || slot;
  return { src: safe ? asset.data : curated || defaultArt(inherited?.fallback || category, theme.illustrated !== false),
    fallback: defaultArt(category, theme.illustrated !== false),
    position: `${focus?.focal_x ?? 50}% ${focus?.focal_y ?? 50}%`,
    source: safe ? asset.source : 'default' };
}
export function artImage(visuals: ModuleVisuals | null | undefined, slot: ArtSlot | null | undefined, fallback: ArtCategory, alt = '') {
  const art = resolveArt(visuals, slot, fallback), image = document.createElement('img');
  image.alt = alt; image.src = art.src; image.style.objectPosition = art.position;
  image.onerror = () => { image.onerror = null; image.src = art.fallback; };
  return image;
}
export function applyTheme(element: HTMLElement, visuals?: ModuleVisuals | null) {
  const theme = themeOf(visuals);
  element.dataset.illustrated = String(theme.illustrated !== false); element.dataset.theme = theme.preset; element.dataset.frame = theme.frame || 'ornate'; element.dataset.font = theme.heading_font || 'serif';
}

/** Portable uploads are decoded/re-encoded locally; no image or model API call. */
export async function prepareUpload(file: File): Promise<VisualAsset> {
  if (!['image/png', 'image/jpeg', 'image/webp'].includes(file.type) || file.size > 8_000_000) throw new Error('请选择 8 MB 以内的 PNG、JPEG 或 WebP 图片。');
  const bitmap = await createImageBitmap(file);
  try {
    if (bitmap.width > 8192 || bitmap.height > 8192) throw new Error('原图边长请不超过 8192 像素。');
    const ratio = Math.min(1, 1200 / Math.max(bitmap.width, bitmap.height));
    const canvas = document.createElement('canvas'); canvas.width = Math.max(1, Math.round(bitmap.width * ratio)); canvas.height = Math.max(1, Math.round(bitmap.height * ratio));
    const ctx = canvas.getContext('2d'); if (!ctx) throw new Error('当前浏览器无法处理图片。');
    ctx.fillStyle = '#e3d6b9'; ctx.fillRect(0, 0, canvas.width, canvas.height); ctx.drawImage(bitmap, 0, 0, canvas.width, canvas.height);
    for (const quality of [.86, .72, .55, .38]) {
      const data = canvas.toDataURL('image/jpeg', quality);
      if (data.length <= 240000) return { name: file.name.slice(0, 120), source: 'upload', data };
    }
    throw new Error('图片细节较多，请先缩小后再上传。');
  } finally { bitmap.close(); }
}

File: app/backend/src/content/visuals.py
"""Portable presentation-only contract. No mechanics or executable theme code."""
import base64
import binascii
import struct
from typing import Literal, Protocol
from pydantic import BaseModel, ConfigDict, Field, model_validator


class VisualDefinition(BaseModel):
    model_config = ConfigDict(extra='forbid')


def image_dimensions(raw: bytes, mime: str) -> tuple[int, int]:
    if mime == 'image/png' and raw[:8] == b'\x89PNG\r\n\x1a\n' and raw[12:16] == b'IHDR' and len(raw) >= 33:
        return struct.unpack('>II', raw[16:24])
    if mime == 'image/jpeg' and raw[:2] == b'\xff\xd8':
        i = 2
        while i + 4 <= len(raw):
            if raw[i] != 255:
                break
            marker = raw[i + 1]
            if marker == 255:
                i += 1
                continue
            if marker in (0xD8, 0xD9) or 0xD0 <= marker <= 0xD7:
                i += 2
                continue
            length = int.from_bytes(raw[i + 2:i + 4], 'big')
            if length < 2 or i + 2 + length > len(raw):
                break
            if marker in (0xC0, 0xC1, 0xC2) and length >= 8:
                h, w = struct.unpack('>HH', raw[i + 5:i + 9])
                return w, h
            i += 2 + length
    raise ValueError('图片内容与格式不符，支持静态 PNG / JPEG。')


class VisualAsset(VisualDefinition):
    name: str = Field(max_length=120)
    source: Literal['upload', 'generated'] = 'upload'
    data: str = Field(max_length=240_000)
    generation_id: str | None = Field(default=None, max_length=100)

    @model_validator(mode='after')
    def raster(self):
        prefix, sep, encoded = self.data.partition(',')
        mime = prefix.removeprefix('data:').removesuffix(';base64')
        if not sep or prefix not in ('data:image/png;base64', 'data:image/jpeg;base64'):
            raise ValueError('素材须为内嵌 PNG / JPEG，不接受网址、SVG 或脚本。')
        try:
            raw = base64.b64decode(encoded, validate=True)
            w, h = image_dimensions(raw, mime)
        except (ValueError, binascii.Error, struct.error) as exc:
            raise ValueError('图片格式无效或内容损坏。') from exc
        if not (0 < w <= 2048 and 0 < h <= 2048):
            raise ValueError('图片边长不能超过 2048 像素。')
        return self


Category = Literal['tavern', 'wilds', 'harbor', 'ruins', 'abstract', 'portrait', 'traveller', 'weapon', 'armor', 'potion', 'item']


class ArtSlot(VisualDefinition):
    builtin: Literal['tavern', 'wilds', 'harbor', 'ruins', 'traveller', 'village-square', 'marcus', 'ayla', 'hooded-merchant', 'dungeon-gate', 'dungeon-passage', 'ancient-temple', 'ancient-vault'] | None = None
    asset_id: str | None = Field(default=None, max_length=80)
    fallback: Category | None = None
    focal_x: int = Field(default=50, ge=0, le=100)
    focal_y: int = Field(default=50, ge=0, le=100)


class ThemePack(VisualDefinition):
    schema_version: Literal[1] = 1
    id: str = Field(default='western-journal', pattern=r'^[a-zA-Z0-9_-]{1,64}$')
    name: str = Field(default='西境旅行志', min_length=1, max_length=80)
    preset: Literal['parchment', 'midnight', 'neutral'] = 'parchment'
    heading_font: Literal['serif', 'sans'] = 'serif'
    frame: Literal['ornate', 'simple'] = 'ornate'
    illustrated: bool = True
    assets: dict[str, VisualAsset] = Field(default_factory=dict, max_length=30)
    defaults: dict[str, ArtSlot] = Field(default_factory=dict, max_length=11)

    @model_validator(mode='after')
    def references(self):
        allowed = {'tavern', 'wilds', 'harbor', 'ruins', 'abstract', 'portrait', 'traveller', 'weapon', 'armor', 'potion', 'item'}
        for key, slot in self.defaults.items():
            if key not in allowed or (slot.asset_id and slot.asset_id not in self.assets):
                raise ValueError('主题默认素材分类或引用无效。')
        if sum(len(a.data) for a in self.assets.values()) > 768_000:
            raise ValueError('主题素材总量请控制在 750 KB 以内。')
        return self


class MapPoint(VisualDefinition):
    x: int = Field(ge=80, le=920)
    y: int = Field(ge=80, le=540)
    label: str | None = Field(default=None, max_length=12)


class ModuleVisuals(VisualDefinition):
    schema_version: Literal[1] = 1
    theme: ThemePack = Field(default_factory=ThemePack)
    assets: dict[str, VisualAsset] = Field(default_factory=dict, max_length=60)
    scenes: dict[str, ArtSlot] = Field(default_factory=dict)
    characters: dict[str, ArtSlot] = Field(default_factory=dict)
    items: dict[str, ArtSlot] = Field(default_factory=dict)
    map: dict[str, ArtSlot] = Field(default_factory=dict)
    map_layout: dict[str, MapPoint] = Field(default_factory=dict)
    player: ArtSlot | None = None
    cover: ArtSlot | None = None

    @model_validator(mode='after')
    def references(self):
        slots = [*self.scenes.values(), *self.characters.values(), *self.items.values(), *self.map.values(), self.player, self.cover]
        for slot in slots:
            if slot and slot.asset_id and slot.asset_id not in self.assets:
                raise ValueError(f'模组素材引用不存在：{slot.asset_id}')
        if sum(len(a.data) for a in [*self.assets.values(), *self.theme.assets.values()]) > 768_000:
            raise ValueError('模组与主题素材合计请控制在 750 KB 以内。')
        return self


# Future provider integration: server-issued quotes, explicit acceptance, then
# an idempotent job. A generation_id in an imported asset is not a paid receipt.
class ImageGenerationRequest(VisualDefinition):
    prompt: str = Field(min_length=1, max_length=4000)
    slot: Category
    theme_id: str = Field(max_length=64)
    count: int = Field(default=1, ge=1, le=4)


class ImageQuote(VisualDefinition):
    id: str
    currency: Literal['CNY'] = 'CNY'
    unit_price_fen: int = Field(ge=0)
    count: int = Field(ge=1, le=4)
    expires_at: str


class ImageJobRequest(VisualDefinition):
    quote_id: str = Field(min_length=1, max_length=100)
    idempotency_key: str = Field(min_length=16, max_length=100)


class ImageJob(VisualDefinition):
    id: str
    status: Literal['queued', 'running', 'succeeded', 'failed']
    assets: list[VisualAsset] = Field(default_factory=list)
    charged_fen: int = Field(default=0, ge=0)


class ImageProvider(Protocol):
    async def quote(self, request: ImageGenerationRequest) -> ImageQuote: ...
    async def submit(self, request: ImageJobRequest) -> ImageJob: ...
    async def status(self, job_id: str) -> ImageJob: ...
